client/driver/docker_progress.go

```diff
@@ -0,0 +1,264 @@
+package driver
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/docker/docker/pkg/jsonmessage"
+	units "github.com/docker/go-units"
+)
+
+const (
+	// dockerPullActivityDeadline is the default value set in the imageProgressManager
+	// when newImageProgressManager is called
+	dockerPullActivityDeadline = 2 * time.Minute
+
+	// dockerImageProgressReportInterval is the default value set in the
+	// imageProgressManager when newImageProgressManager is called
+	dockerImageProgressReportInterval = 10 * time.Second
+)
+
+// layerProgress tracks the state and downloaded bytes of a single layer within
+// a docker image
+type layerProgress struct {
+	id           string
+	status       layerProgressStatus
+	currentBytes int64
+	totalBytes   int64
+}
+
+type layerProgressStatus int
+
+const (
+	layerProgressStatusUnknown layerProgressStatus = iota
+	layerProgressStatusStarting
+	layerProgressStatusWaiting
+	layerProgressStatusDownloading
+	layerProgressStatusVerifying
+	layerProgressStatusDownloaded
+	layerProgressStatusExtracting
+	layerProgressStatusComplete
+	layerProgressStatusExists
+)
+
+func lpsFromString(status string) layerProgressStatus {
+	switch status {
+	case "Pulling fs layer":
+		return layerProgressStatusStarting
+	case "Waiting":
+		return layerProgressStatusWaiting
+	case "Downloading":
+		return layerProgressStatusDownloading
+	case "Verifying Checksum":
+		return layerProgressStatusVerifying
+	case "Download complete":
+		return layerProgressStatusDownloaded
+	case "Extracting":
+		return layerProgressStatusExtracting
+	case "Pull complete":
+		return layerProgressStatusComplete
```

Comment:
This one appears to be defined but not used, is that intentional?

Reply:
Some statuses aren't referenced directly but are used in the comparison on [line 104](https://github.com/hashicorp/nomad/pull/4192/files#diff-b75037f61d70a3fb1ad0d80837e97e86R104). I could condense them all down into the same int, but I figured it would be best to enumerate all of them.